src/Compilers/CSharp/Test/Syntax/Parsing/FileModifierParsingTests.cs

```diff
@@ -0,0 +1,1821 @@
+        [Theory]
+        [InlineData(SyntaxKind.RecordKeyword)]
+        public void FileModifier_04(SyntaxKind typeKeyword)
+        {
+            UsingNode($$"""
+                partial file {{SyntaxFacts.GetText(typeKeyword)}} C { }
+                """);
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxFacts.GetBaseTypeDeclarationKind(typeKeyword));
+                {
+                    N(SyntaxKind.PartialKeyword);
+                    N(SyntaxKind.FileKeyword);
+                    N(typeKeyword);
+                    N(SyntaxKind.IdentifierToken, "C");
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void FileModifier_05()
+        {
+            UsingNode($$"""
+                file partial record struct C { }
+                """);
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.RecordStructDeclaration);
+                {
+                    N(SyntaxKind.FileKeyword);
+                    N(SyntaxKind.PartialKeyword);
+                    N(SyntaxKind.RecordKeyword);
+                    N(SyntaxKind.StructKeyword);
+                    N(SyntaxKind.IdentifierToken, "C");
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        // PROTOTYPE(ft): is it fine that records parse here, but not other type kinds?
+        [Fact]
+        public void FileModifier_06()
+        {
+            UsingNode($$"""
+                partial file record struct C { }
+                """);
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.RecordStructDeclaration);
+                {
+                    N(SyntaxKind.PartialKeyword);
+                    N(SyntaxKind.FileKeyword);
+                    N(SyntaxKind.RecordKeyword);
+                    N(SyntaxKind.StructKeyword);
+                    N(SyntaxKind.IdentifierToken, "C");
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void FileModifier_07()
+        {
+            UsingNode($$"""
+                file partial ref struct C { }
+                """,
+                // (1,14): error CS1585: Member modifier 'ref' must precede the member type and name
+                // file partial ref struct C { }
+                Diagnostic(ErrorCode.ERR_BadModifierLocation, "ref").WithArguments("ref").WithLocation(1, 14));
```

Comment:
This parse and diagnostic are just awful. I know you're not really changing it that much, but this is just wrong.

Reply:
Have filed #61025